tests/core/resource/test_record.py

```diff
@@ -227,6 +227,134 @@ def test_patch_record_returns_updated_fields(self):
         self.assertEquals(self.stored['id'], self.result['id'])
         self.assertEquals(self.result['position'], 10)
 
+    def test_json_patch_add(self):
+        self.resource.request.json = {'data': {'a': 'aaa', 'b': ['bb']}}
+        self.resource.patch()
+        header = self.resource.request.headers
+        header['Content-Type'] = 'application/json-patch+json'
+        self.resource.request.json = [
+            {'op': 'add', 'path': '/data/c', 'value': 'ccc'},
+            {'op': 'add', 'path': '/data/b/1', 'value': 'bbb'},
+            {'op': 'add', 'path': '/data/b/-', 'value': 'bbbb'},
+            {'op': 'add', 'path': '/data/b/0/f/g', 'value': 'fff'},
+        ]
+        result = self.resource.patch()['data']
+        self.assertEqual(result['c'], 'ccc')
+        self.assertEqual(result['b'][1], 'bbb')
+        self.assertEqual(result['b'][2], 'bbbb')
+        self.assertEqual(result['b'][0]['f']['g'], 'fff')
+
+    def test_json_patch_remove(self):
+        self.resource.request.json = {'data': {'a': 'aaa', 'b': ['bb','bbb']}}
+        self.resource.patch()
+        header = self.resource.request.headers
+        header['Content-Type'] = 'application/json-patch+json'
+        self.resource.request.json = [
+            {'op': 'remove', 'path': '/data/a'},
+            {'op': 'remove', 'path': '/data/b/0'},
+        ]
+        result = self.resource.patch()['data']
+        self.assertNotIn('a', result)
+        self.assertEqual(len(result['b']), 1)
+
+        self.resource.request.json = [
+            {'op': 'remove', 'path': '/data/f'},
+        ]
+        self.assertRaises(httpexceptions.HTTPBadRequest, self.resource.patch)
+
+    def test_json_patch_test(self):
+        self.resource.request.json = {'data': {'a': 'aaa', 'b': ['bb','bbb']}}
+        self.resource.patch()
+        header = self.resource.request.headers
+        header['Content-Type'] = 'application/json-patch+json'
+        self.resource.request.json = [
+            {'op': 'test', 'path': '/data/a', 'value': 'aaa'},
+            {'op': 'remove', 'path': '/data/b/1'},
+        ]
+        result = self.resource.patch()['data']
+        self.assertEqual(len(result['b']), 1)
+
+        self.resource.request.json = [
+            {'op': 'test', 'path': '/data/b/0', 'value': 'bbb'},
+            {'op': 'remove', 'path': '/data/b/0'},
+        ]
+        self.assertRaises(httpexceptions.HTTPBadRequest, self.resource.patch)
```

Comment:
I suggest that you keep only *one* group of assertions per test. For example, you could split into two:
* test_json_patch_test_success
* test_json_patch_test_failure

Also, here we don't know if it failed because of the `remove` or the `test`